### PR TITLE
Plane: added aerobatics scripting autotest

### PR DESCRIPTION
**Pull request by tridge**
This will lock in the script working, and also tests a lot of our lua functionality
This PR includes some other required PRs:
- #22163 
- #22171 
- #22153

### CODE REVIEW

File: Tools/autotest/arduplane.py [hunk 1]
```diff
@@ -3958,6 +3958,96 @@ def MidAirDisarmDisallowed(self):
...
+
+        self.context_push()
+        self.set_parameter("TRIK_ENABLE", 1)
+        self.set_rc(7, 1000)
+
+        self.scripting_restart()
+        self.set_parameters({
+            "TRIK1_ID": 72,
+            "RC7_OPTION" : 300,
+            "RC9_OPTION" : 301,
```
Comment (peterbarker) [inline Tools/autotest/arduplane.py:4000]:
```suggestion
            "RC7_OPTION" : 300,  # awesome comment on what this switch option is goes here
            "RC9_OPTION" : 301,  # even more awesome comment goes here
```

File: Tools/autotest/arduplane.py [hunk 2]
```diff
@@ -3958,6 +3958,96 @@ def MidAirDisarmDisallowed(self):
...
+
+        self.wait_text("Trick 1 selected (SuperAirShow)", check_context=True)
+        self.set_rc(7, 2000)
+        self.wait_text("Trick 1 started (SuperAirShow)", check_context=True)
+
+        highest_error = 0
+        while True:
+            m = self.mav.recv_match(type='NAMED_VALUE_FLOAT', blocking=True, timeout=2)
+            if not m:
+                break
```
Comment (peterbarker) [inline Tools/autotest/arduplane.py:4022]:
```suggestion
            m = self.assert_receive_named_value_float("YOUR_NAMED_VALUE_FLOAT_NAME_HERE", timeout=2)
```

You're not filtering to a single named value float here, so if we start sending more random instances (like `AS2`) then we might break this test.

File: Tools/autotest/arduplane.py [hunk 3]
```diff
@@ -3958,6 +3958,96 @@ def MidAirDisarmDisallowed(self):
...
+        self.disarm_vehicle(force=True)
+
+        self.remove_example_script(applet_script)
+        self.remove_example_script(trick72)
+        self.context_pop()
+        self.reboot_sitl()
+        self.remove_message_hook(my_message_hook)
+
+        if ex is not None:
+            raise ex
```
Comment (peterbarker) [inline Tools/autotest/arduplane.py:4037]: Probably want to remove those example scripts after this exception block.

Did you mean to have pretty much everything above this  actually *in* the try/except block?

File: Tools/autotest/common.py [hunk 4]
```diff
@@ -2021,6 +2021,20 @@ def detect_and_handle_reboot(self, old_bootcount, required_bootcount=None, timeo
...
+        arg = mavutil.mavlink.SCRIPTING_CMD_STOP_AND_RESTART
+        self.progress("Restarting Scripting")
+        self.mav.mav.command_int_send(
+            self.sysid_thismav(), 0,
+            mavutil.mavlink.MAV_FRAME_GLOBAL_RELATIVE_ALT_INT,
+            cmd,
+            0, 0,
+            arg,
+            0, 0, 0, 0, 0, 0)
+        self.run_cmd_get_ack(cmd, mavutil.mavlink.MAV_RESULT_ACCEPTED, 5)
```
Comment (peterbarker) [inline Tools/autotest/common.py:2036]:
```suggestion
        self.progress("Restarting Scripting")
        self.run_cmd_int(
            mavutil.mavlink.MAV_CMD_SCRIPTING
            mavutil.mavlink.SCRIPTING_CMD_STOP_AND_RESTART,
            0, 0, 0, 0, 0, 0,
            timeout=5
        )
```

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: Tools/autotest/arduplane.py [hunk 5]
```diff
@@ -3958,6 +3958,96 @@ def MidAirDisarmDisallowed(self):
...
+        tricks = [ "Loop", "HalfReverseCubanEight", "ScaleFigureEight", "Immelmann",
+                   "Split-S", "RollingCircle", "HumptyBump", "HalfCubanEight",
+                   "Upline45", "Downline45", "HalfReverseCubanEight",
+                   "Finishing SuperAirShow!" ]
+        texts = [m.text for m in messages]
+        for t in tricks:
+            if t in texts:
+                self.progress("Completed trick %s" % t)
+            else:
+                raise NotAchievedException("Missing trick %s" % t)
```
[the remark this answers is not in the document]
Comment (tridge) [reply Tools/autotest/arduplane.py]: I want it to be an ordered check, so deliberately didn't use a set

File: Tools/autotest/common.py [hunk 6]
```diff
@@ -2021,6 +2021,20 @@ def detect_and_handle_reboot(self, old_bootcount, required_bootcount=None, timeo
...
+        arg = mavutil.mavlink.SCRIPTING_CMD_STOP_AND_RESTART
+        self.progress("Restarting Scripting")
+        self.mav.mav.command_int_send(
+            self.sysid_thismav(), 0,
+            mavutil.mavlink.MAV_FRAME_GLOBAL_RELATIVE_ALT_INT,
+            cmd,
+            0, 0,
+            arg,
+            0,0,0,0,0,0)
+        self.run_cmd_get_ack(cmd, mavutil.mavlink.MAV_RESULT_ACCEPTED, 5)
```
[the remark this answers is not in the document]
Comment (tridge) [reply Tools/autotest/common.py]: changed, thanks

File: Tools/autotest/common.py (hunk 4 above)
Comment (tridge) [reply Tools/autotest/common.py]: done

File: Tools/autotest/arduplane.py (hunk 2 above)
Comment (tridge) [reply Tools/autotest/arduplane.py:4022]: fixed

File: Tools/autotest/arduplane.py [hunk 7]
```diff
@@ -3958,6 +3958,96 @@ def MidAirDisarmDisallowed(self):
...
+        applet_script = "Aerobatics/FixedWing/plane_aerobatics.lua"
+        trick72 = "Aerobatics/FixedWing/trick72.txt"
+        messages = []
+
+        def my_message_hook(mav, message):
+            if message.get_type() != 'STATUSTEXT':
+                return
+            messages.append(message)
+        self.install_message_hook(my_message_hook)
+
```
[the remark this answers is not in the document]
Comment (tridge) [reply Tools/autotest/arduplane.py]: done

File: Tools/autotest/arduplane.py [hunk 8]
```diff
@@ -3958,6 +3958,96 @@ def MidAirDisarmDisallowed(self):
...
+            [],
+            model=model,
+            defaults_filepath="",
+            wipe=True)
+
+        try:
+            self.install_applet_script(applet_script)
+            self.install_applet_script(trick72)
+            self.reboot_sitl()
+            self.delay_sim_time(10)
```
[the remark this answers is not in the document]
Comment (tridge) [reply Tools/autotest/arduplane.py]: i removed completely, the param set retries works

File: Tools/autotest/arduplane.py (hunk 1 above)
Comment (tridge) [reply Tools/autotest/arduplane.py]: done

File: Tools/autotest/arduplane.py (hunk 3 above)
Comment (tridge) [reply Tools/autotest/arduplane.py]: fixed

File: Tools/autotest/arduplane.py [hunk 9]
```diff
@@ -3958,6 +3958,96 @@ def MidAirDisarmDisallowed(self):
         self.disarm_vehicle(force=True)
         self.reboot_sitl()
 
+    def AerobaticsScripting(self):
+        '''Fixed Wing Aerobatics'''
+        self.start_subtest("Aerobatics Scripting")
+        ex = None
```
[the remark this answers is not in the document]
Comment (tridge) [reply Tools/autotest/arduplane.py]: done

File: Tools/autotest/arduplane.py [hunk 10]
```diff
@@ -3958,6 +3958,96 @@ def MidAirDisarmDisallowed(self):
         self.disarm_vehicle(force=True)
         self.reboot_sitl()
 
+    def AerobaticsScripting(self):
+        '''Fixed Wing Aerobatics'''
+        self.start_subtest("Aerobatics Scripting")
```
[the remark this answers is not in the document]
Comment (tridge) [reply Tools/autotest/arduplane.py]: done

File: Tools/autotest/arduplane.py [hunk 11]
```diff
@@ -3958,6 +3958,96 @@ def MidAirDisarmDisallowed(self):
...
+            wipe=True)
+
+        try:
+            self.install_applet_script(applet_script)
+            self.install_applet_script(trick72)
+            self.reboot_sitl()
+            self.delay_sim_time(10)
+        except Exception as e:
+            self.print_exception_caught(e)
+            ex = e
```
[the remark this answers is not in the document]
Comment (tridge) [reply Tools/autotest/arduplane.py]: fixed

**Review by peterbarker — APPROVED** [review]
Only looked at the autotest.

**Pull request merged**